### PR TITLE
FlatGeobuf format input/output

### PR DESCRIPTION
**Pull request by bjornharrtell**
New take replacing https://github.com/postgis/postgis/pull/583, using bundled FlatBuffers reference C++ implementation via C wrapper.

See http://flatgeobuf.org/ for more on the format.

See for https://lists.osgeo.org/pipermail/postgis-devel/2020-September/028610.html for a discussion in relation to PostGIS.

Proposed input functions:

* ST_TableFromFlatGeobuf(text, text, bytea) RETURNS void
* ST_FromFlatGeobuf(anyelement, bytea) RETURNS setof anyelement

The design is derived from the proposal on how to handle "Various complete table types" at https://stackoverflow.com/questions/11740256/refactor-a-pl-pgsql-function-to-return-the-output-of-various-select-queries.

Proposed output functions is ST_AsFlatGeobuf with similar signature and behavior as ST_AsGeobuf but with the optional parameter to create with spatial index.

#### Must

* [x] CRS
* [x] Datetime
* [x] Extent
* [x] Optional spatial index generation
* [x] Documentation

#### Nice

* [ ] JSON
* [ ] Mixed geometries
* [ ] Additional column metadata

### CODE REVIEW

**Comment by bjornharrtell** [comment]
Ugh.. Seems I randomly get garbage timestamptz roundtrip in the test runs.

**Comment by bjornharrtell** [comment]
This is now feature complete. Note that a large part of the diff is flatbuffers 2.0 headers and generated code from the FlatGeobuf schema, both can be regarded as coming from upstream. Original code for postgis integration is in flatgeobuf.c/h, flatgeobuf_c.cpp/h, geometryreader.cpp/h, geometrywriter.cpp/h, lwgeom_in_flatgeobuf.c and lwgeom_out_flatgeobuf.c.

File: deps/flatgeobuf/Makefile.in [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+# * but WITHOUT ANY WARRANTY; without even the implied warranty of
+# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# * GNU General Public License for more details.
+# *
+# * You should have received a copy of the GNU General Public License
+# * along with PostGIS.  If not, see <http://www.gnu.org/licenses/>.
+# *
+# **********************************************************************
+# *
+# * Copyright 2019 Raúl Marín
```
Comment (Komzpa) [inline deps/flatgeobuf/Makefile.in:21]: mention yourself?

File: deps/flatgeobuf/flatgeobuf_c.h [hunk 2]
```diff
@@ -0,0 +1,108 @@
+#ifndef FLATGEOBUF_C_H
+#define FLATGEOBUF_C_H
```
Comment (Komzpa) [inline deps/flatgeobuf/flatgeobuf_c.h:2]: is this coming from somewhere? I expected flatbuffers to be not tightly coupled with flatgeobuf?

File: deps/flatgeobuf/geometryreader.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
+#ifndef FLATGEOBUF_GEOMETRYREADER_H
+#define FLATGEOBUF_GEOMETRYREADER_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "lwgeom_log.h"
+#include "liblwgeom.h"
```
Comment (Komzpa) [inline deps/flatgeobuf/geometryreader.h:9]: this depends on lwgeom so definitely not third-party dependency.

File: deps/flatgeobuf/packedrtree.cpp [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+    i1 = (i1 | (i1 << 4)) & 0x0F0F0F0F;
+    i1 = (i1 | (i1 << 2)) & 0x33333333;
+    i1 = (i1 | (i1 << 1)) & 0x55555555;
+
+    uint32_t value = ((i1 << 1) | i0);
+
+    return value;
+}
+
+uint32_t hilbert(const NodeItem &r, uint32_t hilbertMax, const double minX, const double minY, const double width, const double height)
```
Comment (Komzpa) [inline deps/flatgeobuf/packedrtree.cpp:130]: we have it a bit differently in core postgis: https://github.com/postgis/postgis/blob/18ce69feba33b13d9fc6122dd594669d9b1bb977/liblwgeom/gbox.c#L796

File: deps/flatgeobuf/packedrtree.cpp [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+        throw std::invalid_argument("Node size must be at least 2");
+    if (_numItems == 0)
+        throw std::invalid_argument("Cannot create empty tree");
+    _nodeSize = std::min(std::max(nodeSize, static_cast<uint16_t>(2)), static_cast<uint16_t>(65535));
+    _levelBounds = generateLevelBounds(_numItems, _nodeSize);
+    _numNodes = _levelBounds.front().second;
+    _nodeItems = new NodeItem[static_cast<size_t>(_numNodes)];
+}
+
+std::vector<std::pair<uint64_t, uint64_t>> PackedRTree::generateLevelBounds(const uint64_t numItems, const uint16_t nodeSize) {
```
Comment (Komzpa) [inline deps/flatgeobuf/packedrtree.cpp:199]: This looks exactly like new gist build logic. @x4m and I have a theory that whole tree will be a lot better if we run picksplit on consecutive pairs of pages, deoverlapping them.

https://github.com/postgis/postgis/blob/f6def67654c25d812446239036cee44812613748/postgis/gserialized_spgist_2d.c#L337

File: deps/flatgeobuf/packedrtree.cpp [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+    return true;
+}
+
+std::vector<double> NodeItem::toVector()
+{
+    return std::vector<double> { minX, minY, maxX, maxY };
+}
+
+// Based on public domain code at https://github.com/rawrunprotected/hilbert_curves
+uint32_t hilbert(uint32_t x, uint32_t y)
```
Comment (Komzpa) [inline deps/flatgeobuf/packedrtree.cpp:81]: 64bit output func is available. helps to not bother with prescaling too much.
https://github.com/postgis/postgis/blob/18ce69feba33b13d9fc6122dd594669d9b1bb977/liblwgeom/lwinline.h#L259

File: postgis/flatgeobuf.h [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PostGIS.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **********************************************************************
+ *
+ * Copyright (C) 2020 Björn Harrtell <bjorn@wololo.org>
```
Comment (Komzpa) [inline postgis/flatgeobuf.h:21]: 2021

File: deps/flatgeobuf/packedrtree.cpp (hunk 4 above)
Comment (bjornharrtell) [reply deps/flatgeobuf/packedrtree.cpp:130]: packedrtree.cpp is from upstream flatgeobuf and I would like to not have to customize that so I hope that duplication is ok.

File: deps/flatgeobuf/packedrtree.cpp (hunk 5 above)
Comment (bjornharrtell) [reply deps/flatgeobuf/packedrtree.cpp:199]: Out of my depth :) This is based on https://github.com/mourner/flatbush (but up side down)

File: deps/flatgeobuf/flatgeobuf_c.h (hunk 2 above)
Comment (bjornharrtell) [reply deps/flatgeobuf/flatgeobuf_c.h:26]: No this is the C API wrapper. I've added a banner.

File: deps/flatgeobuf/geometryreader.h (hunk 3 above)
Comment (bjornharrtell) [reply deps/flatgeobuf/geometryreader.h:33]: Banner added.

File: deps/flatgeobuf/packedrtree.cpp (hunk 6 above)
Comment (bjornharrtell) [reply deps/flatgeobuf/packedrtree.cpp:81]: As mentioned before I wish to keep packedrtree unmodified.

**Comment by robe2** [comment]
Dronie seems to be doing well, but on my local computer, I'm having issues compiling under mingw64.

```
make[2]: Entering directory '/projects/postgis/postgis-git/deps/flatgeobuf'
gcc -std=c++11 -x c++ -I../../liblwgeom -I./ -Iflatbuffers  -IC:/ming64gcc81/projects/geos/rel-3.10w64gcc81/include -IC:/ming64gcc81/projects/proj/rel-7.2.0w64gcc81/include   -IC:/ming64gcc81/projects/protobuf/rel-3.2.0w64gcc81/include   -I/projects/libxml/rel-libxml2-2.9.9w64gcc81/include/libxml2 -I/projects/CGAL/rel-sfcgal-1.3.10w64gcc81/include -IC:/ming64gcc81/projects/json-c/rel-0.12w64gcc81/include/json-c   -IC:/ming64gcc81/projects/pcre/rel-8.33w64gcc81/include   -DNDEBUG -I/projects/postgresql/rel/pg14w64gcc81/include -I/projects/rel-libiconv-1.15w64gcc81/include   -DDLL_EXPORT -DPIC -c -o flatgeobuf_c.o flatgeobuf_c.cpp
In file included from feature_generated.h:7,
                 from flatgeobuf_c.cpp:26:
flatbuffers/flatbuffers.h:20:10: fatal error: flatbuffers/base.h: No such file or directory
 #include "flatbuffers/base.h"
          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [Makefile:49: flatgeobuf_c.o] Error 1
make[2]: Leaving directory '/projects/postgis/postgis-git/deps/flatgeobuf'
make[1]: *** [Makefile:213: ../deps/flatgeobuf/libflatgeobuf.la] Error 2
make[1]: Leaving directory '/projects/postgis/postgis-git/postgis'
make: *** [GNUmakefile:22: all] Error 1

```

I thought maybe it was because flatbuffers/base.h was not declared as a header in your Makefile.in, but putting it in did not help.

**Comment by robe2** [comment]
Okay I managed to get it to compile but getting this regression failure which at a glance appears to be a timezone difference - mine is registering in Boston Time where as the expect result is in UTC.

```
 ./regress/core/flatgeobuf .. failed (diff expected obtained: /projects/postgis/tmp/3.2.0dev_pg14_geos3.10_gdal3.2.0w64/test_18_diff)
-----------------------------------------------------------------------------
--- ./regress/core/flatgeobuf_expected  2021-09-19 14:27:18.691664200 -0400
+++ /projects/postgis/tmp/3.2.0dev_pg14_geos3.10_gdal3.2.0w64/test_18_out       2021-09-20 01:17:58.214182100 -0400
@@ -19,4 +19,4 @@
 P1|0|POINT(1 2)
 P1|1|POINT(3 4)
 --- Attribute roundtrips ---
-A1|0||t|1|2|3|4|1.2|1.3|2016-06-23 03:44:52.134125+00|hello
+A1|0||t|1|2|3|4|1.2|1.3|2016-06-22 23:44:52.134125-04|hello
-----------------------------------------------------------------------------

```

To get it to compile I had to make two changes to get past the include issue, have no clue why but it worked:
in:
flatbuffers/stl_emulation.h

changed:

```
#include "flatbuffers/base.h"
```
to 
```
#include "base.h"
```

In  flatbuffers/flatbuffers.h

changed 
```
#include "flatbuffers/base.h"
#include "flatbuffers/stl_emulation.h"
```

to
```
#include "base.h"
#include "stl_emulation.h"
```

**Comment by bjornharrtell** [comment]
Yeah I half expected the timezone issue to come up. Not sure how to best handle that.

The include path issue must have something to do with that it does not like current path includes in a subdir which is strange.

**Comment by robe2** [comment]
> 
> 
> Yeah I half expected the timezone issue to come up. Not sure how to best handle that.
> 
> The include path issue must have something to do with that it does not like current path includes in a subdir which is strange.

Yah and why I didn't need to touch the other includes that had flatbuffers/ in them.  I thought maybe the recursive includes was tripping it up somehow.  It didn't seem to have the issue with wagyu though granted wagyu has .hpp and <> instead of quoted. Trying that as I recall didn't help (putting in <>)

As far as timezone, can you set the timezone at the beginning of the test.  Something like:

```
SET timezone to UTC;
```
I'll give that a try to see it if works.

**Comment by robe2** [comment]
Looks all good on my end now.

**Comment by bjornharrtell** [comment]
Merged as of https://git.osgeo.org/gitea/postgis/postgis/commit/3ae2c584400b6a04c59f70b0ca28e6a63331d9e1.

**Pull request closed without merging**